### PR TITLE
Adding code lens for BUILD files

### PR DESCRIPTION
**Pull request by smukherj1**
Gives a convenient way to directly launch builds or tests by clicking on the annotations in the BUILD files

### CODE REVIEW

File: src/codelens/provider.ts [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+/**
+ * Search for the path to the directory that has the Bazel WORKSPACE file for the given file.
+ * Returns the path to the directory with the Bazel WORKSPACE if found. Returns undefined otherwise.
+ * If multiple directories along the path to the file has files called "WORKSPACE", the lowest path
+ * is returned.
+ * @param fsPath Path to a file in a Bazel workspace
+ */
+function getBazelWorkspaceFolder(fsPath: string): string | undefined {
+  var b, d: string
```
[the remark this answers is not in the document]
Comment (smukherj1) [reply src/codelens/provider.ts]: Done

File: src/codelens/provider.ts [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+  /**
+   * Takes the result of a Bazel query for targets defined in a package and returns a list of
+   * CodeLens for the BUILD file in that package
+   * @param bazelWorkspaceDirectory The bazel workspace directory
+   * @param queryResult Result of the bazel query requesting targets in a package
+   */
+  addCodeLens(bazelWorkspaceDirectory: string, queryResult: QueryResult): vscode.CodeLens[] {
+    let result = [];
+
+    for (let i = 0; i < queryResult.rules.length; i++) {
```
[the remark this answers is not in the document]
Comment (smukherj1) [reply src/codelens/provider.ts]: Done

File: src/codelens/provider.ts [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+      let r = queryResult.rules[i];
+      // Source location in the BUILD file for this target
+      let loc = r.location;
+      // Fully qualified name of the target
+      let target = r.name;
+      let rc = r.ruleClass;
+      var cmd: vscode.Command;
+      if (rc.endsWith("_test")) {
+        cmd = {
+          title: "Test " + target,
```
[the remark this answers is not in the document]
Comment (smukherj1) [reply src/codelens/provider.ts]: Done

File: src/codelens/provider.ts [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { BazelQuery, QueryResult } from "../bazel/query"
+import * as vscode from "vscode";
+import * as path from "path";
+import * as fs from "fs";
+import { BazelTest, BazelBuild, BazelCommandAdapter, BazelCommandArgs } from "../bazel/commands";
```
[the remark this answers is not in the document]
Comment (smukherj1) [reply src/codelens/provider.ts]: Done

File: src/codelens/provider.ts [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    if (relDirWithDoc === ".") {
+      relDirWithDoc = "";
+    }
+    // Turn the relative path into a package label
+    let pkg = "//" + relDirWithDoc;
+    let queryResult = await new BazelQuery(w,
+      "'kind(rule, " + pkg + ":all)'", []).runAndParse();
+    return this.addCodeLens(w, queryResult);
+  }
+}
```
[the remark this answers is not in the document]
Comment (smukherj1) [reply src/codelens/provider.ts]: Done

File: src/extension/extension.ts [hunk 6]
```diff
@@ -31,8 +32,16 @@ export function activate(context: vscode.ExtensionContext) {
     // Commands
     vscode.commands.registerCommand("bazel.buildTarget", bazelBuildTarget),
     vscode.commands.registerCommand("bazel.buildTargetWithDebugging",
-                                    bazelBuildTargetWithDebugging),
+      bazelBuildTargetWithDebugging),
```
[the remark this answers is not in the document]
Comment (smukherj1) [reply src/extension/extension.ts]: I'm using VS Code as my editor and I use whatever formatter VS Code defaults to

[the remark this answers is not in the document]
Comment (allevato) [reply src/extension/extension.ts]: Hmm, mine should be the same. I'll investigate later.

**Pull request merged**